### PR TITLE
Allow for routing Kibana and Produce requests to K8s

### PR DESCRIPTION
**Pull request by bugbounce**
Refactor producer router interface

Add producer router for k8s

Fixes for multiple producer case

Add ability to redirect K8s Kibana

### CODE REVIEW

File: router/producer_router.go [hunk 1]
```diff
@@ -66,69 +66,142 @@ func (p *producerRouter) Server() *http.Server {
...
+		req.Header.Get(AppNameHeaderName),
+	)
+	span.SetTag("app-group", profile.ClusterName)
+
+	//Collect all results and errors for all the produce endpoints.
+	var produceResults []*pb.ProduceResult
+	var produceErrors []error
+
+	if req.Body != nil {
+		reqBody, _ = ioutil.ReadAll(req.Body)
```
Comment (ekarisky) [inline router/producer_router.go:96]: please check if we can return if req.Body is nil

File: router/producer_router.go [hunk 2]
```diff
@@ -167,35 +258,69 @@ func (p *producerRouter) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+		}
+	}
+
+	// Return failure if any of the produce attempts had errors
+	if responseMsg.Len() > 0 {
+
+		w.Write(responseMsg.Bytes())
+
+		if len(errorCodes) > 0 {
+			w.WriteHeader(errorCodes[0])
```
Comment (ekarisky) [inline router/producer_router.go:311]: can we modify existing error handling function `onRpcError(w, err)`?

File: router/producer_router_test.go [hunk 3]
```diff
@@ -263,6 +340,85 @@ func TestProducerRouter_WithAppGroupSecret(t *testing.T) {
...
+	testPayload = sampleRawTimberCollection()
+	req, _ = http.NewRequest(http.MethodGet, "http://localhost/produce_batch", bytes.NewBuffer(testPayload))
+	req.Header.Add("X-App-Secret", "some-secret")
+	resp = RecordResponse(router.ServeHTTP, req)
+
+	FatalIfWrongResponseStatus(t, resp, http.StatusOK)
+	FatalIfWrongResponseBody(t, resp, "")
+}
+
 func NewTestSuccessfulProducer(ctrl *gomock.Controller, marketUrl string, host string, producerPort int, consulPort int) ProducerRouter {
```
Comment (ekarisky) [inline router/producer_router_test.go:422]: please add postfix `_LXC` for old test, as we will have `K8s` and `DoubleWrite` test

File: router/kibana_router_test.go [hunk 4]
```diff
@@ -106,6 +106,29 @@ func TestKibanaRouter(t *testing.T) {
 	FatalIfWrongResponseBody(t, resp, "some-target")
 }
 
+func TestKibanaRouter_k8s(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (galuhest) [reply router/kibana_router_test.go]: fixed

File: router/producer_router_test.go (hunk 3 above)
Comment (galuhest) [reply router/producer_router_test.go]: fixed

File: router/producer_router.go (hunk 2 above)
Comment (galuhest) [reply router/producer_router.go]: this is done

File: router/producer_router.go (hunk 1 above)
Comment (galuhest) [reply router/producer_router.go:94]: we tried to do this, but it's failing because the response code is wrong/different

**Review by ekarisky — APPROVED** [review]

**Pull request merged**